### PR TITLE
Presentation: Add an option to apply content modifiers to nested content

### PR DESCRIPTION
**Pull request by medikornov**
Closes https://github.com/iTwin/itwinjs-core/issues/3786

imodel-native: https://github.com/iTwin/imodel-native/tree/presentation/content-modifiers-on-nested-content

### CODE REVIEW

File: presentation/common/src/presentation-common/rules/content/modifiers/ContentModifier.ts [hunk 1]
```diff
@@ -65,4 +65,11 @@ export interface ContentModifier extends RuleBase, ContentModifiersList {
    * if this attribute is not specified.
    */
   class?: SingleSchemaClassSpecification;
+
+  /**
+   * Specifies whether [calculatedProperties]($docs/presentation/content/CalculatedPropertiesSpecification.md) and
+   * [relatedProperties]($docs/presentation/content/RelatedPropertiesSpecification.md) specifications should also be applied on
```
Comment (grigasp) [inline presentation/common/src/presentation-common/rules/content/modifiers/ContentModifier.ts:71]:
```suggestion
   * Specifies whether [`calculatedProperties`]($docs/presentation/content/CalculatedPropertiesSpecification.md) and
   * [`relatedProperties`]($docs/presentation/content/RelatedPropertiesSpecification.md) specifications should also be applied on
```

File: full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts [hunk 2]
```diff
@@ -176,6 +176,95 @@ describe("Learning Snippets", () => {
...
+                    schemaName: "BisCore",
+                    className: "SpatialViewDefinitionUsesModelSelector",
+                  },
+                  direction: "Backward",
+                  targetClass: { schemaName: "BisCore", className: "SpatialViewDefinition" },
+                },
+              }],
+            }],
+          }, {
+            // apply related and calculated properties on nested properties
```
Comment (grigasp) [inline full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts:204]:
```suggestion
            // add a content modifier for `SpatialViewDefinition` elements
```

File: full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts [hunk 3]
```diff
@@ -176,6 +176,95 @@ describe("Learning Snippets", () => {
...
+                },
+                direction: "Forward",
+                targetClass: { schemaName: "BisCore", className: "DisplayStyle" },
+              },
+            }],
+            calculatedProperties: [{
+              label: "Calculated for SpatialViewDefinition",
+              value: "this.Extents",
+            }],
+            applyOnNestedContent: true,
```
Comment (grigasp) [inline full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts:221]:
```suggestion
            // set `applyOnNestedContent: true` to get this modifier applied when `SpatialViewDefinition` content is loaded indirectly
            applyOnNestedContent: true,
```

**Review by grigasp — CHANGES_REQUESTED** [review]
TODOs:
- [ ] Add to `/docs/presentation/content/ContentModifier.md`.
- [ ] Add a section to `NextVersion.md`
- [ ] Update our supplemental rulesets at `/presentation/backend/assets/supplemental-presentation-rules` to make use of this new attribute.

File: full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts [hunk 4]
```diff
@@ -176,6 +176,95 @@ describe("Learning Snippets", () => {
...
+              // load content for given input instances
+              specType: "SelectedNodeInstances",
+              relatedProperties: [{
+                propertiesSource: {
+                  relationship: {
+                    schemaName: "BisCore",
+                    className: "SpatialViewDefinitionUsesModelSelector",
+                  },
+                  direction: "Backward",
+                  targetClass: { schemaName: "BisCore", className: "SpatialViewDefinition" },
```
[the remark this answers is not in the document]
Comment (medikornov) [reply full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts]: fixed

File: full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts [hunk 5]
```diff
@@ -176,6 +176,95 @@ describe("Learning Snippets", () => {
...
+                  { label: "User Label" },
+                ],
+              },
+            ],
+          },
+        ]);
+        const spatialViewDefinition = content!.contentSet[0].displayValues[getFieldByLabel(content!.descriptor.fields, "Spatial View Definition").name] as DisplayValuesArray;
+        expect(spatialViewDefinition.length).to.eq(1);
+        const spatialViewDefinitionProperties = (spatialViewDefinition[0] as DisplayValuesMap).displayValues as DisplayValuesMap;
+        expect(spatialViewDefinitionProperties[getFieldByLabel(content!.descriptor.fields, "Calculated for SpatialViewDefinition").name]).to.eq("5.9371887957409877,3.2424499696297646,40");
```
[the remark this answers is not in the document]
Comment (medikornov) [reply full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts]: fixed?

File: presentation/common/src/presentation-common/rules/content/modifiers/ContentModifier.ts (hunk 1 above)
Comment (medikornov) [reply presentation/common/src/presentation-common/rules/content/modifiers/ContentModifier.ts]: fixed

File: full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts (hunk 2 above)
Comment (medikornov) [reply full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts]: fixed

File: full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts (hunk 3 above)
Comment (medikornov) [reply full-stack-tests/presentation/src/learning-snippets/content/rules/ContentModifier.test.ts:220]: fixed

File: docs/changehistory/NextVersion.md [hunk 6]
```diff
@@ -100,6 +100,10 @@ Previously, the backend would return a "timeout" status if creating the response
...
 
+### Use content modifiers on nested content
+
+Previously, the calculated and related properties defined in the content modifier were only applied to displayed instances, but not to related instances. In most cases, when showing related instance properties, we also want to apply them to all nested content. Now, the content modifier has an attribute `applyOnNestedContent` which indicates if the modifier should be used on nested content. See [content modifier](../presentation/content/ContentModifier.md) for more details.
```
Comment (grigasp) [inline docs/changehistory/NextVersion.md:105]:
```suggestion
Previously, the [calculated](../presentation/content/ContentModifier.md#attribute-calculatedproperties) and [related properties](../presentation/content/ContentModifier.md#attribute-relatedproperties) defined in [content modifiers](../presentation/content/ContentModifier.md) were only applied on directly loaded instances' content. Occasionally, there is a need to request calculated and/or related properties to be loaded for specific instances in all situations, no matter if their content is loaded directly or indirectly. Previously that was only possible by chaining [related properties](../presentation/content/ContentModifier.md#attribute-relatedproperties) and [nested related properties](../presentation/content/RelatedPropertiesSpecification.md#attribute-nestedrelatedproperties) attributes. Now the content modifier rule has an attribute [`applyOnNestedContent`](../presentation/content/ContentModifier.md#attribute-applyonnestedcontent) which indicates if the modifier should be used on nested content. This removes the need to have duplicate [related properties specifications](../presentation/content/RelatedPropertiesSpecification.md) in those situations.
```

File: presentation/common/src/presentation-common/rules/content/modifiers/ContentModifier.ts [hunk 7]
```diff
@@ -65,4 +65,11 @@ export interface ContentModifier extends RuleBase, ContentModifiersList {
    * if this attribute is not specified.
    */
   class?: SingleSchemaClassSpecification;
+
+  /**
+   * Specifies whether [`calculatedProperties`]($docs/presentation/content/CalculatedPropertiesSpecification.md) and
+   * [`relatedProperties`]($docs/presentation/content/RelatedPropertiesSpecification.md) specifications should also be applied on
+   * [nested content]($docs/presentation/content/Terminology.md#nested-content).
+   */
+  applyOnNestedContent?: boolean;
```
[the remark this answers is not in the document]
Comment (medikornov) [reply presentation/common/src/presentation-common/rules/content/modifiers/ContentModifier.ts:75]: fixed

**Review by grigasp — APPROVED** [review]
Need `extract-api` and `rush change`, but otherwise looks good

**Pull request merged**